Add global prop to Tracks events for user role

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/woocommerce/issues/52402

This PR adds global properties, `role`, `can_install_plugins`, `can_activate_plugins` and `can_manage_woo`, to Track data to capture the current user's role. By including this property, we aim to improve insights into how different user roles interact with WooCommerce features, helping to address potential permission-related issues better. The role property is derived from the first item in the current user's roles array, following WordPress conventions.


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Set up a new store, make sure to opt-in to Tracks.
1. Trigger a Tracks event by open up browser console and run `wcTracks.recordEvent('test_role', {})`
1. Go to networks tab and filter for `test_role`
1. Confirm that each Tracks event includes `role`, `can_install_plugins`, `can_activate_plugins` and `can_manage_woo` properties.
1. Go to `WooCommerce > Home` and click on an inbox note action button to trigger server-side event.
1. Go to `WooCommerce > Status > Logs` and find the `tracks-*` logs.
1. Confirm that the `role`, `can_install_plugins`, `can_activate_plugins` and `can_manage_woo` properties are included in the Tracks event.
1. Repeat the above steps with different user roles such as Shop Manager to ensure the role property is correctly captured.


<img width="1147" alt="Screenshot 2024-11-06 at 11 06 15" src="https://github.com/user-attachments/assets/900a5ca2-9a04-4a8a-8515-a34d92c80e8f">


<img width="1147" alt="Screenshot 2024-11-04 at 16 05 38" src="https://github.com/user-attachments/assets/27bbbe0e-c04c-43b3-8342-224c7efb6949">


### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
Add global prop to Tracks events for user role

chihsuan committed:
Add changelog

chihsuan committed:
Add unit tests

rjchow reviewed (commented):
Hmm I wonder if this is sufficient - to my understanding 'roles' are a means of packaging up permissions, but what we actually check for (and causes errors) are permissions. 

Should we only track roles or also include a few permissions that we typically check against, such as `install_plugins`, `activate_plugins`, and `manage_woocommerce`?

chihsuan commented:
> I wonder if this is sufficient - to my understanding 'roles' are a means of packaging up permissions, but what we actually check for (and causes errors) are permissions.
> 
> Should we only track roles or also include a few permissions that we typically check against, such as `install_plugins`, `activate_plugins`, and `manage_woocommerce`?

That's a good point! @rjchow 👍  I agree that we should track permissions. I think maybe we could track both roles and permissions since it doesn't cost much. Knowing a user's role would help us understand user behavior and feature needs by role and improve security and role-based access. What do you think? 

cc @adrianduffell

adrianduffell commented:
I'm happy to track permissions, in addition or instead of roles. I was a bit hesitant originally because I think we would need to use booleans for each permission, e.g. `has_install_plugins`, `has_ activate_plugins`, `has_manage_woocommerce` to allow them to be used in Tracks funnels.

chihsuan commented:
> I'm happy to track permissions, in addition or instead of roles. I was a bit hesitant originally because I think we would need to use booleans for each permission, e.g. `has_install_plugins`, `has_activate_plugins`, `has_manage_woocommerce` to allow them to be used in Tracks funnels.

Thanks for the feedback. I'll make the changes to add permissions tracking.

chihsuan committed:
Enhance user role details in WC_Tracks to include capabilities for installing, activating plugins, and managing WooCommerce

chihsuan commented:
I've updated the PR to track both roles and permissions. 5596e7db21a7e06ffeede66740c73d39b287d320

I also removed the `_` prefix from the properties as it looks like we only use that for service properties.

psealock commented on plugins/woocommerce/includes/tracks/class-wc-tracks.php:
```diff
@@ -67,6 +67,21 @@ public static function get_server_details() {
+	 */
+	public static function get_role_details( $user ) {
+		return array(
+			'role'                   => ! empty( $user->roles ) ? $user->roles[0] : '',
```
Is this the default logic? Will there be a case where user can have multiple roles? If so, what order are they in? Unit tests around this would be good too, just to be clear about the desired logic.

chihsuan replied:
Adrian [found that](https://github.com/woocommerce/woocommerce/issues/52402) WordPress only provides a UI for giving a user a single role, but it is programmatically possible to give a user multiple roles.

We decided to keep it simple so that we only consider the first role in the list of roles for a user. I will add unit tests to cover this!

chihsuan replied:
Updated unit tests in 01bf835ec2c52d01e43c639047b2323131b62c4a.

chihsuan committed:
Update unit tests

rjchow reviewed (approved):
tested well, thank you!

psealock reviewed (approved):
Thanks for the changes @chihsuan 👍🏽
